Voxelisation and MC association for edep-sim reco application

jback08 opened the pull request.

lhwhitehead commented on include/PandoraInterface.h:
```diff
@@ -99,8 +230,150 @@ inline Parameters::Parameters() :
+    m_invDir({0, 0, 0}),
+    m_sign({0, 0, 0})
+{
+    float dx = m_dir.GetX();
+    float dy = m_dir.GetY();
+    float dz = m_dir.GetZ();
```
Can be const?

lhwhitehead commented on include/PandoraInterface.h:
```diff
@@ -99,8 +230,150 @@ inline Parameters::Parameters() :
+    m_binExtent(binWidths.GetMagnitude()),
+    m_nBins({0, 0, 0})
+{
+    long NxBins = long((top.GetX() - bottom.GetX()) / binWidths.GetX());
+    long NyBins = long((top.GetY() - bottom.GetY()) / binWidths.GetY());
+    long NzBins = long((top.GetZ() - bottom.GetZ()) / binWidths.GetZ());
```
Can be const? Also best to static cast instead of use a C-style one?

lhwhitehead commented on include/PandoraInterface.h:
```diff
@@ -99,8 +230,150 @@ inline Parameters::Parameters() :
+{
+    // Bin widths should always be non-zero. Need to use long integers to avoid
+    // integer overflow for the total bin, which can be larger than 2^31
+    long xBin = getBinIndex(point.GetX(), m_bottom.GetX(), m_binWidths.GetX(), m_nBins[0]);
+    long yBin = getBinIndex(point.GetY(), m_bottom.GetY(), m_binWidths.GetY(), m_nBins[1]);
+    long zBin = getBinIndex(point.GetZ(), m_bottom.GetZ(), m_binWidths.GetZ(), m_nBins[2]);
+
+    long totBin = (zBin * m_nBins[1] + yBin) * m_nBins[0] + xBin;
```
consts?

lhwhitehead commented on test/PandoraInterface.cxx:
```diff
@@ -163,36 +164,35 @@ void CreateGeometry(const Parameters &parameters, const Pandora *const pPrimaryP
-    double dx = pBox->GetDX() * mm2cm; // Note these are the half widths
-    double dy = pBox->GetDY() * mm2cm;
-    double dz = pBox->GetDZ() * mm2cm;
+    double dx = pBox->GetDX() * m_mm2cm; // Note these are the half widths
+    double dy = pBox->GetDY() * m_mm2cm;
+    double dz = pBox->GetDZ() * m_mm2cm;
```
const?

lhwhitehead commented on test/PandoraInterface.cxx:
```diff
@@ -336,14 +368,462 @@ void ProcessEvents(const Parameters &parameters, const Pandora *const pPrimaryPa
+    else if (!inStart && !inStop)
+    {
+        //std::cout << "Case 2: Start and end points are outside boundary" << std::endl;
+        bool ok = grid.intersect(ray, t0, t1);
```
const?

lhwhitehead commented on test/PandoraInterface.cxx:
```diff
@@ -336,14 +368,462 @@ void ProcessEvents(const Parameters &parameters, const Pandora *const pPrimaryPa
+    {
+        //std::cout << "Case 3: Start inside boundary" << std::endl;
+        point1 = start;
+        bool ok = grid.intersect(ray, t0, t1);
```
const

lhwhitehead commented on test/PandoraInterface.cxx:
```diff
@@ -336,14 +368,462 @@ void ProcessEvents(const Parameters &parameters, const Pandora *const pPrimaryPa
+    {
+        //std::cout << "Case 4: End inside boundary" << std::endl;
+        point2 = stop;
+        bool ok = grid.intersect(ray, t0, t1);
```
const

lhwhitehead commented on test/PandoraInterface.cxx:
```diff
@@ -336,14 +368,462 @@ void ProcessEvents(const Parameters &parameters, const Pandora *const pPrimaryPa
+bool LArBox::intersect(const LArRay &ray, double &t0, double &t1) const
+{
+    // Brian Smits ray-box algorithm with improvements from Amy Williams et al
+    double x(ray.m_origin.GetX());
+    double y(ray.m_origin.GetY());
+
+    double invDirX(ray.m_invDir.GetX());
+    double invDirY(ray.m_invDir.GetY());
```
const

jback08 commented:
Presentation showing voxelisation and 1st pass cosmic-ray-based reconstruction for single muon tracks is available here:
https://indico.fnal.gov/event/50101

lhwhitehead commented on test/PandoraInterface.cxx:
```diff
@@ -336,14 +368,462 @@ void ProcessEvents(const Parameters &parameters, const Pandora *const pPrimaryPa
+{
+    bool result(false);
+
+    float x = point.GetX();
+    float y = point.GetY();
+    float z = point.GetZ();
```
const (sorry)

lhwhitehead commented on test/PandoraInterface.cxx:
```diff
@@ -336,14 +368,462 @@ void ProcessEvents(const Parameters &parameters, const Pandora *const pPrimaryPa
+        tMax = tyMax;
+    }
+
+    double z(ray.m_origin.GetZ());
+    double invDirZ(ray.m_invDir.GetZ());
```
const

lhwhitehead commented on test/PandoraInterface.cxx:
```diff
@@ -336,14 +368,462 @@ void ProcessEvents(const Parameters &parameters, const Pandora *const pPrimaryPa
+    std::vector<LArVoxel> mergedVoxels;
+
+    const int nVoxels = voxelList.size();
+    std::vector<int> processed(nVoxels, 0);
```
strictly std::vector<bool>?

jback08 replied:
Changed to use booleans. I prefer to keep this as a vector to preserve the (initial) voxel ordering. The loop over voxel pairs (i and j) gets progressively faster, since all earlier voxels will be processed beforehand.

lhwhitehead commented on test/PandoraInterface.cxx:
```diff
@@ -336,14 +368,462 @@ void ProcessEvents(const Parameters &parameters, const Pandora *const pPrimaryPa
+    // Ray direction will be the same, but update starting point
+    ray.updateOrigin(point1);
+
+    const double epsilon(1e-3);
```
Is this an arbitrary value?

jback08 replied:
Yes, it is just a small distance (but not too small) to move along the path.

lhwhitehead commented on test/PandoraInterface.cxx:
```diff
@@ -336,14 +368,462 @@ void ProcessEvents(const Parameters &parameters, const Pandora *const pPrimaryPa
+        // Voxel box
+        const LArBox vBox(voxBot, voxTop);
+        // Get ray intersections with this box
+        bool result = vBox.intersect(ray, t0, t1);
+
+        if (!result)
```
Result only used on this one line, so just put the `vBox.intersect` call in the if statement

jback08 replied:
Decided to move this to be part of the if statement as suggested, since it also updates t0 and t1.

lhwhitehead commented on test/PandoraInterface.cxx:
```diff
@@ -336,14 +368,462 @@ void ProcessEvents(const Parameters &parameters, const Pandora *const pPrimaryPa
+        }
+
+        // Voxel extent = intersection path difference
+        double dL(t1 - t0);
+        // For the first path length, use the distance from the
+        // starting ray point to the 2nd intersection t1
+        if (loop == 0)
+        {
+            dL = t1;
+        }
```
Could simplify into a ternary operator? Though perhaps it would lose readability.

jback08 replied:
Prefer to keep things as they are for readability.

PandoraPFA commented on test/PandoraInterface.cxx:
```diff
@@ -336,14 +368,464 @@ void ProcessEvents(const Parameters &parameters, const Pandora *const pPrimaryPa
+
+        // Grid 3d bin containing this point; 4th element is the total bin number
+        std::array<long, 4> gridBins = grid.getBinIndices(voxelPoint);
+        long voxelID = gridBins[3];
```
const

PandoraPFA commented on test/PandoraInterface.cxx:
```diff
@@ -336,14 +368,464 @@ void ProcessEvents(const Parameters &parameters, const Pandora *const pPrimaryPa
+        std::array<long, 4> gridBins = grid.getBinIndices(voxelPoint);
+        long voxelID = gridBins[3];
+
+        long xBin = gridBins[0];
```
const

PandoraPFA commented on test/PandoraInterface.cxx:
```diff
@@ -336,14 +368,464 @@ void ProcessEvents(const Parameters &parameters, const Pandora *const pPrimaryPa
+        }
+
+        // Voxel energy using path length fraction w.r.t hit length
+        const double voxelEnergy = g4HitEnergy * dL / hitLength;
```
hitLength zero checked anywhere? double vs. float.

jback08 replied:
Checks are made on hit length and hit energy to be > 0. Changed to float.

PandoraPFA commented on test/PandoraInterface.cxx:
```diff
@@ -336,14 +368,464 @@ void ProcessEvents(const Parameters &parameters, const Pandora *const pPrimaryPa
+
+        // Store voxel object in vector
+        const LArVoxel voxel(voxelID, voxelEnergy, voxBot, trackID);
+        currentVoxelList.push_back(voxel);
```
Or emplace_back?

jback08 replied:
`push_back` changed to `emplace_back` here and a few other places elsewhere.

PandoraPFA commented on test/PandoraInterface.cxx:
```diff
@@ -336,14 +368,464 @@ void ProcessEvents(const Parameters &parameters, const Pandora *const pPrimaryPa
+    ray.updateOrigin(point1);
+
+    // Small position shift to ensure we move along the ray path
+    const double epsilon(1e-3);
```
How does the value of this impact the efficiency? Any concerns at all about efficiency?

jback08 replied:
Don't think we have tried to find an optimal value for this. We just need a small path length change (but not too small) so that we continue along the hit segment "trajectory" to find the next (different) voxel bin.

jback08 replied:
This path shift has been added as the constant value `m_voxelPathShift` in `Parameters`. Allows us the possibility to include this as an optional runtime parameter within `ParseCommandLine` if we discover efficiency issues related to this.

PandoraPFA commented on test/PandoraInterface.cxx:
```diff
@@ -336,14 +368,464 @@ void ProcessEvents(const Parameters &parameters, const Pandora *const pPrimaryPa
+
+    return currentVoxelList;
+}
+
```
Two dividing lines, as moving to implementation of member functions from a different/new class.

PandoraPFA commented on test/PandoraInterface.cxx:
```diff
@@ -336,14 +368,464 @@ void ProcessEvents(const Parameters &parameters, const Pandora *const pPrimaryPa
+
+bool LArBox::intersect(const LArRay &ray, double &t0, double &t1) const
+{
+    // Brian Smits ray-box algorithm with improvements from Amy Williams et al
```
Any good reference to add here? Was this comment from source (under a compatible licence) reused from elsewhere? As per above, should acknowledge source and compatible licence.

jback08 replied:
https://www.scratchapixel.com/lessons/3d-basic-rendering/minimal-ray-tracer-rendering-simple-shapes/ray-box-intersection
"An Efficient and Robust Ray–Box Intersection Algorithm". Amy Williams et al. 2004: https://doi.org/10.1145/1198555.1198748
"Efficiency Issues for Ray Tracing", Brian Smits 1998: https://doi.org/10.1080/10867651.1998.10487488.

PandoraPFA replied:
Great, maybe some small comment with useful references (DoIs shouldn't go out of date), and then due diligence where any code is reused.

PandoraPFA commented on test/PandoraInterface.cxx:
```diff
@@ -336,14 +368,464 @@ void ProcessEvents(const Parameters &parameters, const Pandora *const pPrimaryPa
+    t1 = tMax;
+    return true;
+}
+
```
Dividing line.

PandoraPFA commented on test/PandoraInterface.cxx:
```diff
@@ -336,14 +368,464 @@ void ProcessEvents(const Parameters &parameters, const Pandora *const pPrimaryPa
+
+bool LArBox::inside(const pandora::CartesianVector &point) const
+{
+    bool result(false);
```
remove `result` and just return appropriately below.

PandoraPFA commented on test/PandoraInterface.cxx:
```diff
@@ -336,14 +368,464 @@ void ProcessEvents(const Parameters &parameters, const Pandora *const pPrimaryPa
+
+std::vector<LArVoxel> MergeSameVoxels(const std::vector<LArVoxel> &voxelList)
+{
+
```
Empty line - this kind of cosmetic point should be addressed by clang format.

PandoraPFA commented on test/PandoraInterface.cxx:
```diff
@@ -336,14 +368,464 @@ void ProcessEvents(const Parameters &parameters, const Pandora *const pPrimaryPa
+            continue;
+        }
+
+        LArVoxel voxel1 = voxelList[i];
```
const?

jback08 replied:
No, this can't be const since its energy can be updated to be the sum of all of the other voxels with the same ID, which is then stored in the merged list, avoiding the need to create another voxel object.

PandoraPFA commented on test/PandoraInterface.cxx:
```diff
@@ -336,14 +368,464 @@ void ProcessEvents(const Parameters &parameters, const Pandora *const pPrimaryPa
+        }
+
+        LArVoxel voxel1 = voxelList[i];
+        double voxE1 = voxel1.m_energyInVoxel;
```
const?

PandoraPFA commented on settings/PandoraSettings_EDepReco.xml:
```diff
@@ -14,7 +14,10 @@
+      <MCParticleListNames>Input</MCParticleListNames>
+        <!--Suppress neutrons (PDG=2112) with E < 0.95 GeV-->
+        <SuppressMCParticles>2112:0.95</SuppressMCParticles>
+	<CaloHitListNames>CaloHitListU CaloHitListV CaloHitListW</CaloHitListNames>
```
Check indentation and no tab characters.

PandoraPFA reviewed (commented):
Many thanks @jback08 and @afm1g15 - I've just worked through the diffs and added any thoughts that occurred. Please see the attached review comments. Some can be ignored, others are hopefully worth some thought.

I think we should apply the clang format file, reference any important sources used for any voxelisation code, and generally think about consistency of treatment between e.g. floats vs. doubles, ints vs. unsigned ints vs. longs. It sounds like longs are a requirement, so can use that as a starting point for consistency.

Other than that, can't see any issues immediately, and it seems (from recent slides/examples) that things are working nicely. Thanks again!

jback08 commented on include/PandoraInterface.h:
```diff
@@ -20,15 +21,20 @@ class Pandora;
-     *  @brief Default constructor
-     */
+   *  @brief Default constructor
+   */
```
This file was formatted using clang-format v7.0.0 with a link in the `LArReco` directory pointing to the `.clang-format` file in `PandoraPFA`, with: `clang-format -i -style=file PandoraInterface.cxx`

jback08 replied:
I understand that occasionally clang format does not correctly indent doxygen comments, but have tried to manually correct this when needed.

jback08 commented on include/PandoraInterface.h:
```diff
@@ -43,8 +49,104 @@ class Parameters
+    bool m_use3D;     ///< Create 3D LArCaloHits
+    bool m_useLArTPC; ///< Create LArTPC LArCaloHits with u,v,w views
+
+    double m_voxelWidth; ///< Voxel box width (cm)
```
No, we can use float. So I gather that floats are generally preferred throughout the Pandora code (for memory management)?

PandoraPFA replied:
Yeah, we don't have a rigid policy, but a general preference where the single float precision is clearly sufficient. Switch to doubles for calculations where the precision really matters, etc.

jback08 replied:
The intersection code needs double precision, but essentially everything else can use floats.

jback08 commented on include/PandoraInterface.h:
```diff
@@ -43,8 +49,104 @@ class Parameters
+
+    LArGrid(const LArGrid &rhs);
+
+    std::array<long, 4> getBinIndices(const pandora::CartesianVector &point) const;
```
Long is needed for the total bin number, since we can have values that exceed 2^31.

jback08 commented on include/PandoraInterface.h:
```diff
@@ -99,8 +230,150 @@ inline Parameters::Parameters() :
+
+inline std::array<long, 4> LArGrid::getBinIndices(const pandora::CartesianVector &point) const
+{
+    // Bin widths should always be non-zero. Need to use long integers to avoid
```
We can say long types are required in the `LArVoxel` constructor, which is near the top of the header file.

jback08 commented on include/PandoraInterface.h:
```diff
@@ -20,15 +21,20 @@ class Pandora;
 
+typedef std::map<int, double> MCParticleEnergyMap;
+
+double m_mm2cm{0.1};
```
Could move this to the `Parameters` class.

jback08 commented on include/PandoraInterface.h:
```diff
@@ -43,8 +49,104 @@ class Parameters
+{
+public:
+    /**
+   *  @brief Default constructor
```
VoxelID needs to be long, since the total bin number are be larger than 2^31.

jback08 commented on test/PandoraInterface.cxx:
```diff
@@ -336,14 +368,464 @@ void ProcessEvents(const Parameters &parameters, const Pandora *const pPrimaryPa
+
+std::vector<LArVoxel> MakeVoxels(const TG4HitSegment &g4Hit, const LArGrid &grid)
+{
+    std::vector<LArVoxel> currentVoxelList;
```
Yes, we can include 2 references. The code is based on (GPLv3, compatible with Pandora):
https://www.scratchapixel.com/lessons/3d-basic-rendering/minimal-ray-tracer-rendering-simple-shapes/ray-box-intersection

Alex's first implementation is based on MIT-licensed (GPLv3 compatible) code from https://github.com/chenel/larcv2/tree/edepsim-formattruth/larcv/app/Supera, which also references the above ray-box-intersection URL. The main changes I made were to simplify the intersection logic and to organise the functions and methods into smaller classes.

AndyChappell commented on test/PandoraInterface.cxx:
```diff
@@ -257,40 +271,58 @@ void ProcessEvents(const Parameters &parameters, const Pandora *const pPrimaryPa
-                caloHitParameters.m_mipEquivalentEnergy = energy;
-                caloHitParameters.m_electromagneticEnergy = energy;
-                caloHitParameters.m_hadronicEnergy = energy;
+                caloHitParameters.m_inputEnergy = voxelE;
+                caloHitParameters.m_mipEquivalentEnergy = voxelE;
+                caloHitParameters.m_electromagneticEnergy = voxelE;
+                caloHitParameters.m_hadronicEnergy = voxelE;
```
What are the units of energy here?

jback08 replied:
Energy is GeV, which is converted from the G4 hit energy in MeV within `MakeVoxels()`.

AndyChappell commented on test/PandoraInterface.cxx:
```diff
@@ -336,14 +368,464 @@ void ProcessEvents(const Parameters &parameters, const Pandora *const pPrimaryPa
+            // Should not reach here, but set sensible values just in case
+            mcParticleParameters.m_vertex = pandora::CartesianVector(0, 0, 0);
+            mcParticleParameters.m_endpoint = pandora::CartesianVector(0, 0, 0);
+            mcParticleParameters.m_process = 0;
```
lar_content::MC_PROC_UNKNOWN

jback08 commented on test/PandoraInterface.cxx:
```diff
@@ -336,14 +368,464 @@ void ProcessEvents(const Parameters &parameters, const Pandora *const pPrimaryPa
+
+            } // end voxel loop
+
+            // The voxelisation only works with ArgonCube
```
Some comments have been added here to mention that ArgonCube geometry parameters are assumed and this needs to be generalised to work for any detector.

jback08 commented on test/PandoraInterface.cxx:
```diff
@@ -336,14 +368,464 @@ void ProcessEvents(const Parameters &parameters, const Pandora *const pPrimaryPa
+    }
+
+    // Define ray trajectory (this checks dirMag >= epsilon)
+    const CartesianVector dirNorm = dir.GetUnitVector();
```
The magnitude of the direction vector is checked to be > floating epsilon before this line is reached.

jback08 commented on test/PandoraInterface.cxx:
```diff
@@ -336,14 +368,464 @@ void ProcessEvents(const Parameters &parameters, const Pandora *const pPrimaryPa
+    else if (!inStart && !inStop)
+    {
+        //std::cout << "Case 2: Start and end points are outside boundary" << std::endl;
+        const bool ok = grid.intersect(ray, t0, t1);
```
These booleans have been removed and the functions called within their respective if() statement.

jback08 commented:
OK, I think I have addressed essentially all of the coding style recommendations and suggestions made so far: header file structure, function and class separation lines, doxygen comments, typedefs, removal of braces for single line conditional statements, etc. Have also added references (URLs, license types and papers) for the voxelisation and intersection source code.

The intersection code and related functions needs double precision, but all other variables are set to be float. The voxel total bin needs to be a long integer since it can be larger than 2^31; the other bin indices are long as well to ensure consistent long arithmetic.

AndyChappell commented on test/PandoraInterface.cxx:
```diff
@@ -257,40 +276,56 @@ void ProcessEvents(const Parameters &parameters, const Pandora *const pPrimaryPa
+            std::cout << "Produced " << mergedVoxels.size() << " merged voxels from " << voxelList.size() << " voxels." << std::endl;
+
+            // Loop over the voxels and make them into caloHits
+            for (int i = 0; i < mergedVoxels.size(); i++)
+            {
+                const LArVoxel voxel = mergedVoxels[i];
```
Looks to me like you can dispense with the loop counter for a range-based for loop of the form
for (const LArVoxel : mergeVoxels)

AndyChappell commented on test/PandoraInterface.cxx:
```diff
@@ -163,36 +160,36 @@ void CreateGeometry(const Parameters &parameters, const Pandora *const pPrimaryP
-    Double_t level3[3] = {0.0, 0.0, 0.0};
+    // TODO: Check if this is really needed, as it doesn't seem to change anything
+    double level3[3] = {0.0, 0.0, 0.0};
     currentnode->LocalToMasterVect(origin, level3);
-    Double_t level2[3] = {0.0, 0.0, 0.0};
+    double level2[3] = {0.0, 0.0, 0.0};
     currentnode->LocalToMasterVect(level3, level2);
-    Double_t level1[3] = {0.0, 0.0, 0.0};
+    double level1[3] = {0.0, 0.0, 0.0};
```
Is the idea here that the local and intermediate coordinates here come from boxes, within boxes, within boxes, and that these steps promote the local coordinates to the global coordinates, or is something other structure involved here? That these translations apparently don't do anything suggest I have the wrong model.

jback08 replied:
We only need 1 local-to-global transform so the code has been updated accordingly.

AndyChappell reviewed (approved):
Just one trivial update and a query on what the "levels" represent, but otherwise this looks ready to merge to me.
